Mathlib/Topology/Spectral/Hom.lean

```diff
@@ -0,0 +1,225 @@
+compact open set is compact open.
+
+## Main declarations
+
+* `IsSpectralMap`: Predicate for a map to be spectral.
+* `SpectralMap`: Bundled spectral maps.
+* `SpectralMapClass`: Typeclass for a type to be a type of spectral maps.
+
+## TODO
+
+Once we have `SpectralSpace`, `IsSpectralMap` should move to `topology.spectral.basic`.
+-/
+
+
+open Function OrderDual
+
+variable {F α β γ δ : Type _}
+
+section Unbundled
+
+variable [TopologicalSpace α] [TopologicalSpace β] [TopologicalSpace γ] {f : α → β} {s : Set β}
+
+/-- A function between topological spaces is spectral if it is continuous and the preimage of every
+compact open set is compact open. -/
+structure IsSpectralMap (f : α → β) extends Continuous f : Prop where
+  /-- A function between topological spaces is spectral if it is continuous and the preimage of every
+compact open set is compact open. -/
+  isCompact_preimage_of_isOpen ⦃s : Set β⦄ : IsOpen s → IsCompact s → IsCompact (f ⁻¹' s)
+#align is_spectral_map IsSpectralMap
+
+theorem IsCompact.preimage_of_isOpen (hf : IsSpectralMap f) (h₀ : IsCompact s) (h₁ : IsOpen s) :
+    IsCompact (f ⁻¹' s) :=
+  hf.isCompact_preimage_of_isOpen h₁ h₀
+#align is_compact.preimage_of_is_open IsCompact.preimage_of_isOpen
+
+theorem IsSpectralMap.continuous {f : α → β} (hf : IsSpectralMap f) : Continuous f :=
+  hf.toContinuous
+#align is_spectral_map.continuous IsSpectralMap.continuous
+
+theorem isSpectralMap_id : IsSpectralMap (@id α) :=
+  ⟨continuous_id, fun _s _ => id⟩
+#align is_spectral_map_id isSpectralMap_id
+
+theorem IsSpectralMap.comp {f : β → γ} {g : α → β} (hf : IsSpectralMap f) (hg : IsSpectralMap g) :
+    IsSpectralMap (f ∘ g) :=
+  ⟨hf.continuous.comp hg.continuous, fun _s hs₀ hs₁ =>
+    ((hs₁.preimage_of_isOpen hf hs₀).preimage_of_isOpen hg) (hs₀.preimage hf.continuous)⟩
+#align is_spectral_map.comp IsSpectralMap.comp
+
+end Unbundled
+
+/-- The type of spectral maps from `α` to `β`. -/
+structure SpectralMap (α β : Type _) [TopologicalSpace α] [TopologicalSpace β] where
+  /-- function between topological spaces-/
+  toFun : α → β
+  /-- proof that `toFun` is a spectral map-/
+  spectral' : IsSpectralMap toFun
+#align spectral_map SpectralMap
+
+section
+
+/-- `SpectralMapClass F α β` states that `F` is a type of spectral maps.
+
+You should extend this class when you extend `SpectralMap`. -/
+class SpectralMapClass (F : Type _) (α β : outParam <| Type _) [TopologicalSpace α]
+  [TopologicalSpace β] extends FunLike F α fun _ => β where
+  /-- statement that `F` is a type of spectal maps-/
+  map_spectral (f : F) : IsSpectralMap f
+#align spectral_map_class SpectralMapClass
+
+end
+
+export SpectralMapClass (map_spectral)
+
+attribute [simp] map_spectral
+
+-- See note [lower instance priority]
+instance (priority := 100) SpectralMapClass.toContinuousMapClass [TopologicalSpace α]
+    [TopologicalSpace β] [SpectralMapClass F α β] : ContinuousMapClass F α β :=
```

Comment:
This is a limitation of the way Lean 4 handles `outParam`s. Basically, we can't guarantee `[TopologicalSpace α]` is tried after the `[SpectralMapClass F α β]` tells Lean what `α` and `β` are supposed to be. The solution is often to make the parameters depending on `outParam`s into implicits:
```suggestion
instance (priority := 100) SpectralMapClass.toContinuousMapClass {_ : TopologicalSpace α}
    {_ : TopologicalSpace β} [SpectralMapClass F α β] : ContinuousMapClass F α β :=
```
I expect this should work because `SpectralMapClass` is also defined on topological spaces, so the required instances can be taken from the `SpectralMapClass` instance we find.

Reply:
Thanks -- saw the `outParam` fix..